### PR TITLE
Update and expand UDF docs

### PR DESCRIPTION
**Pull request by mosabua**
## Description

Main updates are to rebrand from "SQL routines" to "User-defined functions (UDFs)" and "SQL user-defined functions", rejig all the content so it is an appropriate base, and then add "Python user-defined functions".

More details in the commit messages since there is a lot happening. 

Note to reviewers: Review per commit and potentially build locally to see rendered output. When rewording paragraphs and such wrapping was also always fixed/updated.

## Additional context and related issues

#24378 

## Release notes

Note sure yet .. we might want to add a release note even though this should not break anything or have any user facing context itself.. just so its clear that we are rebranding.

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: docs/src/main/sphinx/udf/function.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ FUNCTION name ( [ parameter_name data_type [, ...] ] )
   [ CALLED ON NULL INPUT ]
   [ SECURITY { DEFINER | INVOKER } ]
   [ COMMENT description]
+  [ WITH properties AS block]
   statements
```
Comment (electrum) [inline docs/src/main/sphinx/udf/function.md:15]: This is the new syntax: either SQL routine statements or a function definition as a dollar-quoted string:
```
{ statements | AS definition }

File: docs/src/main/sphinx/udf/function.md [hunk 2]
```diff
@@ -58,6 +59,10 @@ The `COMMENT` characteristic can be used to provide information about the
 function to other users as `description`. The information is accessible with
 [](/sql/show-functions).
 
+Use the `WITH properties AS block` section to define the `handler` property for
+a [](/udf/python) to specify the Python method to invoke as the function. The
+Python code in `block` is encapsulated within `$$`.
+
 The body of the UDF can either be a simple single `RETURN` statement with an
```
Comment (electrum) [inline docs/src/main/sphinx/udf/function.md:66]: This needs to be split to account for the different syntaxes of SQL vs other languages. Maybe something like

> For SQL UDFs, the body can be ...

> For UDFs in other languages, the definition is a dollar-quoted string ...

File: docs/src/main/sphinx/udf/introduction.md [hunk 3]
```diff
@@ -14,8 +15,9 @@ plugin. Details are available in the [developer guide](/develop/functions).
 (udf-declaration)=
 ## UDF declaration
 
-Declare the UDF with a [](/udf/function) keyword using the supported statements
-for [](/udf/sql).
+Declare the UDF with the SQL [](/udf/function) keyword, additional parameters
+such as `RETURNS` or `LANGUAGE` and use the supported statements for [](/udf/sql) or
```
Comment (electrum) [inline docs/src/main/sphinx/udf/introduction.md:19]: This sentence is a bit hard to read. What if we simplify and combine with below

```markdown
Declare the UDF using the [](/udf/function) declaration. [...]

File: docs/src/main/sphinx/udf/python.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+The following snippet shows pseudo-code:
+
+```text
+  FUNCTION python_udf_name(input_parameter data_type)
+    RETURNS result_data_type
+    LANGUAGE PYTHON
+    WITH (handler = 'python_function')
+    AS
+    $$
```
Comment (electrum) [inline docs/src/main/sphinx/udf/python.md:35]: I'd prefer to use `AS $$` on one line for all the documentation. The other way has too much vertical whitespace.

File: docs/src/main/sphinx/udf/python.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+Set the language to `PYTHON` to override the default `SQL` for [](/udf/sql).
+The Python code is enclosed with ``$$` and must use valid formatting.
+
+```text
+WITH
+  FUNCTION doubleup(x integer)
+    RETURNS integer
+    LANGUAGE PYTHON
+    WITH (handler = 'malzwei')
```
Comment (electrum) [inline docs/src/main/sphinx/udf/python.md:55]: Is this a made up word? Can we choose something more understandable here?

File: docs/src/main/sphinx/udf/python.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+:hidden:
+
+/udf/python/examples
+```
+
+(python-udf-lang)=
+## Python language details
+
+The Trino Python UDF integrations uses Python 3.13.0 in a sandboxed environment.
+Python code runs within a WebAssembley (WASM) runtime within the Java virtual
```
Comment (electrum) [inline docs/src/main/sphinx/udf/python.md:80]: Typo "WebAssembly"

File: docs/src/main/sphinx/udf/python.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+Python language rules including indents must be observed.
+
+Python UDFs therefore only have access to the Python language and core libraries
+included in the sandboxed runtime. Access to external resources with network or
+file system operations is not supported. Usage of other Python libraries as well
+as command line tools or package managers is not supported.
+
+The following libraries are explicitly removed from the runtime and therefore
+not available witin a Python UDF:
```
Comment (electrum) [inline docs/src/main/sphinx/udf/python.md:91]: Typo "within"

File: docs/src/main/sphinx/udf/python.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+* `_*_support*`
+* `_pyrepl`
```
Comment (electrum) [inline docs/src/main/sphinx/udf/python.md:94]: Remove, these are internal

File: docs/src/main/sphinx/udf/python.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+* `bdb`
+* `concurrent`
+* `curses`
+* `ensurepip`
+* `doctest*`
```
Comment (electrum) [inline docs/src/main/sphinx/udf/python.md:99]: Remove all the `*` as that is a filesystem thing and not the Python package name

**Review by electrum — COMMENTED** [review]
Looks great!

File: docs/src/main/sphinx/udf/python.md (hunk 5 above)
Comment (mosabua) [reply docs/src/main/sphinx/udf/python.md:55]: it means timestwo in german .. I will change ;-)

File: docs/src/main/sphinx/udf/introduction.md [hunk 10]
```diff
@@ -1,16 +1,25 @@
...
 
+:::{note}
+User-defined functions can alternatively be written in Java and deployed as a
+plugin. Details are available in the [developer guide](/develop/functions).
+:::
+
+(udf-declaration)=
+## UDF declaration
+
+Declare the UDF with a [](/udf/function) keyword using the supported statements
```
[the remark this answers is not in the document]
Comment (mosabua) [reply docs/src/main/sphinx/udf/introduction.md]: sure .. its sort of duplicate

File: docs/src/main/sphinx/udf/introduction.md [hunk 11]
```diff
@@ -94,75 +105,11 @@ Use the [](/connector/memory) in a catalog for simple storing and
...
-nested blocks and labels.
-
 ## Recommendations
 
 Processing UDFs can potentially be resource intensive on the cluster in
 terms of memory and processing. Take the following considerations into account
-when writing and running SQL UDFs:
+when writing and running UDFs:
 
 * Some checks for the runtime behavior of UDF are in place. For example,
```
[the remark this answers is not in the document]
Comment (mosabua) [reply docs/src/main/sphinx/udf/introduction.md:116]: Reworded to be correct along the overall query processing context.

File: docs/src/main/sphinx/udf/function.md [hunk 12]
```diff
@@ -31,7 +32,7 @@ The `type` value after the `RETURNS` keyword identifies the [data
 type](/language/types) of the UDF output.
 
 The optional `LANGUAGE` characteristic identifies the language used for the UDF
-definition with `language`. Only `SQL` is supported.
+definition with `language`. Only `SQL` or `PYTHON` are supported.
```
[the remark this answers is not in the document]
Comment (mosabua) [reply docs/src/main/sphinx/udf/function.md:35]: updated

File: docs/src/main/sphinx/udf/introduction.md (hunk 3 above)
Comment (mosabua) [reply docs/src/main/sphinx/udf/introduction.md]: Done

File: docs/src/main/sphinx/udf/python.md (hunk 4 above)
Comment (mosabua) [reply docs/src/main/sphinx/udf/python.md:35]: Done

**Pull request merged**